### ISSUE TITLE
Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer@13352000] has been already registered for type: class hzcmd.model.person.PersonExternalizable java.lang.IllegalStateException

### ISSUE THREAD

**Issue by Danny-Hazelcast** [hazelcast/hazelcast#18408]
Failed http://jenkins.hazelcast.com/view/kill/job/kill-sql/73/console
```
15:40:02 /home/ec2-user/workspace/kill-sql/4.2-SNAPSHOT/2021_03_15-15_35_08/sql/bit-map Failed
15:40:02 fail HzClient3HZ put-mapBak1Obj hzcmd.map.PutRandomStylePerson threadId=0 com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable' [0m
15:40:03 http://3.87.100.12/~/home/ec2-user/workspace/kill-sql/4.2-SNAPSHOT/2021_03_15-15_35_08/sql/bit-map
```

but very next run passed

Padded http://jenkins.hazelcast.com/view/kill/job/kill-sql/74/console
```
16:31:22 /home/ec2-user/workspace/kill-sql/4.2-SNAPSHOT/2021_03_15-16_26_18/sql/bit-map Passed
16:31:22 http://3.87.100.12/~/home/ec2-user/workspace/kill-sql/4.2-SNAPSHOT/2021_03_15-16_26_18/sql/bit-map
```

java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer@13352000] has been already registered for type: class ......

could there be some kind of race condition hear ?

```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:178)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:139)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:127)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putInternal(ClientMapProxy.java:525)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.put(ClientMapProxy.java:266)
	at hzcmd.map.PutRandomStylePerson.timeStep(PutRandomStylePerson.java:12)
	at remote.bench.marker.RollMarker.flatOut(RollMarker.java:72)
	at remote.bench.marker.RollMarker.bench(RollMarker.java:59)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer@13352000] has been already registered for type: class hzcmd.model.person.PersonExternalizable
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.safeRegister(AbstractSerializationService.java:441)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.lookupJavaSerializer(AbstractSerializationService.java:581)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:508)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:167)
	... 17 more
```

**Comment by sancar** [comment]
```
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer@13352000] has been already registered for type: class hzcmd.model.person.PersonExternalizable
```

This is very odd. From what I understand from the code, an Externalizable class can not be registered with `JavaSerializer` . In the type map,  for `PersonExternalizable`, we should have `ExternalizableSerializer` instead.

**Comment by sancar** [comment]
These are the classes that is used in the test. 
https://github.com/hazelcast/hzCmd-bench/tree/master/src/main/java/hzcmd/model/person
They are put randomly to the map. 
https://github.com/hazelcast/hzCmd-bench/blob/master/src/main/java/hzcmd/map/PutRandomStylePerson.java

**Comment by Danny-Hazelcast** [comment]
this seams to be a very rare issue,  1 have hit only 1 failing run,    all other times, the tests passes

**Comment by Danny-Hazelcast** [comment]
http://jenkins.hazelcast.com/view/kill/job/kill-x/62/console
sh hz-bench-run-jenkins-kill hz/stable/hd/pool/hd-idx 0 4 M4C4


 /disk1/workspace/kill-x/5.0-SNAPSHOT/2021_04_07-09_59_53/hd-idx Failed
 fail HzClient2HZ fn hzcmd.distributed.PutRandPerson threadId=1 
 com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'


HazelcastClient 5.0-SNAPSHOT (20210406 - e762a54, 58e1a14) is STARTING

```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:178)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:139)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:127)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putInternal(ClientMapProxy.java:525)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.put(ClientMapProxy.java:266)
	at hzcmd.distributed.PutRandPerson.timeStep(PutRandPerson.java:21)
	at remote.bench.marker.RollMarker.flatOut(RollMarker.java:72)
	at remote.bench.marker.RollMarker.bench(RollMarker.java:59)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$ExternalizableSerializer@748986b5] has been already registered for type: class hzcmd.model.person.PersonExternalizable
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.safeRegister(AbstractSerializationService.java:441)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.registerFromSuperType(AbstractSerializationService.java:463)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.lookupCustomSerializer(AbstractSerializationService.java:548)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:500)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:167)
	... 17 more
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:178)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:139)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:127)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putInternal(ClientMapProxy.java:525)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.put(ClientMapProxy.java:266)
	at hzcmd.distributed.PutRandPerson.timeStep(PutRandPerson.java:21)
	at remote.bench.marker.RollMarker.flatOut(RollMarker.java:72)
	at remote.bench.marker.RollMarker.bench(RollMarker.java:59)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$ExternalizableSerializer@748986b5] has been already registered for type: class hzcmd.model.person.PersonExternalizable
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.safeRegister(AbstractSerializationService.java:441)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.registerFromSuperType(AbstractSerializationService.java:463)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.lookupCustomSerializer(AbstractSerializationService.java:548)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:500)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:167)
	... 17 more
```


```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:178)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:139)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:127)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putInternal(ClientMapProxy.java:525)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.put(ClientMapProxy.java:266)
	at hzcmd.distributed.PutRandPerson.timeStep(PutRandPerson.java:21)
	at remote.bench.marker.RollMarker.flatOut(RollMarker.java:72)
	at remote.bench.marker.RollMarker.bench(RollMarker.java:59)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$ExternalizableSerializer@748986b5] has been already registered for type: class hzcmd.model.person.PersonExternalizable
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.safeRegister(AbstractSerializationService.java:441)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.registerFromSuperType(AbstractSerializationService.java:463)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.lookupCustomSerializer(AbstractSerializationService.java:548)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:500)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:167)
	... 17 more
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'hzcmd.model.person.PersonExternalizable'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:178)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:154)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:139)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:127)
	at com.hazelcast.client.impl.spi.ClientProxy.toData(ClientProxy.java:76)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.putInternal(ClientMapProxy.java:525)
	at com.hazelcast.client.impl.proxy.ClientMapProxy.put(ClientMapProxy.java:266)
	at hzcmd.distributed.PutRandPerson.timeStep(PutRandPerson.java:21)
	at remote.bench.marker.RollMarker.flatOut(RollMarker.java:72)
	at remote.bench.marker.RollMarker.bench(RollMarker.java:59)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Serializer[com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$ExternalizableSerializer@748986b5] has been already registered for type: class hzcmd.model.person.PersonExternalizable
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.safeRegister(AbstractSerializationService.java:441)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.registerFromSuperType(AbstractSerializationService.java:463)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.lookupCustomSerializer(AbstractSerializationService.java:548)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:500)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:167)
	... 17 more
```

**Comment by Danny-Hazelcast** [comment]
reliable reproducer for this issue 

http://jenkins.hazelcast.com/view/stable/job/stable-x1/133/console
hz-bench-run-jenkins-stable hz/stress/issue-18408


```
#!/usr/bin/env bash
set -e
for i in {1..100}; do
  hz go hzcmd.distributed.Create count=2 maps=mapBak1HD
  hz go hzcmd.distributed.AddIndex fields=salary.lastName.firstName.age.personId.count
  hz go hzcmd.distributed.PutRandPerson3 threads=16 keyDomain=1 duration=1
  hz bounce
done
```

https://github.com/hazelcast/hzCmd-bench/blob/zeta/src/main/java/hzcmd/distributed/PutRandPerson3.java

only fails when  PersonExternalizable  extends  PersonSerializable

fails on 4.0 4.1 4.2 and 5.0-SNAPSHOT

**Comment by mmedenjak** [comment]
@sancar can you check out the reproducer mentioned above?

**Comment by sancar** [comment]
> only fails when PersonExternalizable extends PersonSerializable

This is the important part for me. In this case, we have a class `PersonExternalizable` and its serializer can change depend on who is put first. 

if order is as follows:
map.put(1, new PersonExternalizable())
map.put(1, new PersonSerializable())

PersonExternalizable will be serialized as `Externalizable` 

And if it is as follows:
map.put(1, new PersonSerializable())
map.put(1, new PersonExternalizable())

PersonExternalizable will be serialized as `Serializable` 

And if you are unlucky enough with multiple threads(I think there needs to be at least 3 threads in motion), then you can get the `already registered ... exception`
Even if we forget about the exception, the order deciding the serializer is problematic. And this is caused by user's setup. I want to say this is user's mistake. 
Basically, a class should not have a super class which meant to be serialized via a different serializer.

 Note that this same problem can happen if a class implementing IdentifiedDataSerializable interface extends a class implementing Portable interface as well. 
 
 @Danny-Hazelcast I know this issue is around for a while and you invested too much but I want to close this issue without a fix.
 What do you think ? 
I haven't realized they are extending each other. I want to say that in the commit I have checked they were not extending but my memory can fail me. It could be the case that I missed it, sorry if it is the case.

**Comment by Danny-Hazelcast** [comment]
if there is nothing that can be done to fix the unlucky race condition where the already registered exception is thrown
then we can closes this issue ?

**Comment by sancar** [comment]
There does not seem to have a solution considering the options we have. Closing the issue.
